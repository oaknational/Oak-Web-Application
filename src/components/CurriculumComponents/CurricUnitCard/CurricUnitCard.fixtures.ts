import { createUnit } from "@/fixtures/curriculum/unit";
import { createUnitOption } from "@/fixtures/curriculum/unitOption";

export const unitWithOptions = createUnit({
  slug: "something-nice",
  unit_options: [
    createUnitOption({
      slug: "really-nice",
    }),
    createUnitOption({
      slug: "kind-of-nice",
    }),
  ],
});

export const unitWithoutOptions = createUnit({
  slug: "something-nice",
  unit_options: [],
});
