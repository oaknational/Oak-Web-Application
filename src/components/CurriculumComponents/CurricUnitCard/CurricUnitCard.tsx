import {
  OakBox,
  OakFlex,
  OakIcon,
  OakSpan,
  OakTypography,
} from "@oaknational/oak-components";
import { VisuallyHidden } from "react-aria";
import styled from "styled-components";
import Link from "next/link";
import { ReactNode } from "react";

import FocusIndicator from "../OakComponentsKitchen/FocusIndicator";

import OutlineHeading from "@/components/SharedComponents/OutlineHeading";
import { TagFunctional } from "@/components/SharedComponents/TagFunctional";
import { HTMLButtonProps } from "@/components/SharedComponents/Button/common";
import { Unit } from "@/utils/curriculum/types";

const StyledLink = styled(Link)<HTMLButtonProps & { $isHighlighted: boolean }>`
  width: 100%;
  height: 100%;
  overflow: hidden;
  background: none;
  border: none;
  padding: 0;
  margin: 0;
  display: block;
  text-align: left;
  outline: none;
  cursor: pointer;
  background: ${({ $isHighlighted }) => ($isHighlighted ? "black" : "white")};
  &:hover {
    background: ${({ $isHighlighted }) =>
      $isHighlighted ? "#575757" : "#F2F2F2"};
  }
`;

type CurricUnitCardProps = {
  unit: Unit | Unit["unit_options"][number];
  index: number;
  isHighlighted: boolean;
  href: string;
  onClick?: () => void;
  additional: ReactNode;
};
export default function CurricUnitCard(props: CurricUnitCardProps) {
  const { href, isHighlighted, unit, index, onClick } = props;
  const isUnitOption = "unit_options" in unit;
  return (
    <FocusIndicator
      $height={"100%"}
      $width={"100%"}
      $borderRadius={"border-radius-s"}
      $overflow={"hidden"}
      disableMouseHover={isHighlighted}
    >
      <StyledLink
        href={href}
        $isHighlighted={isHighlighted}
        data-testid="unit-info-link"
        shallow={true}
        scroll={false}
        replace={true}
        onClick={onClick}
      >
        <OakFlex
          $pv={"inner-padding-s"}
          $ph={"inner-padding-m"}
          $height={"100%"}
          $width={"100%"}
          $color={isHighlighted ? "white" : "black"}
          data-testid={isHighlighted ? "highlighted-unit-card" : "unit-card"}
          $flexDirection={"column"}
        >
          <OakFlex $flexDirection={"column"} $gap={"space-between-xs"}>
            <OutlineHeading tag={"div"} $font={"heading-5"} $fontSize={24}>
              <span aria-hidden={true}>{index + 1}</span>
            </OutlineHeading>
            <OakSpan $font={"heading-7"}>{unit.title}</OakSpan>
            {isUnitOption && unit.unit_options.length > 1 && (
              <OakBox
                $zIndex={"neutral"}
                data-testid="options-tag"
                $position={"relative"}
              >
                <TagFunctional
                  color="lavender"
                  text={`${unit.unit_options.length} unit options`}
                />
              </OakBox>
            )}
            {props.additional && <OakBox>{props.additional}</OakBox>}
          </OakFlex>

          <OakFlex
            $flexDirection={"row"}
            $justifyContent={"flex-end"}
            $mt={"space-between-s"}
            $flexGrow={1}
            $alignItems={"flex-end"}
          >
            <OakFlex $alignItems={"center"} $gap={"space-between-sssx"}>
              <OakTypography $font={"heading-7"}>Unit info</OakTypography>

              <OakIcon
                $width="all-spacing-6"
                $height="all-spacing-6"
                $colorFilter={isHighlighted ? "white" : "black"}
                alt=""
                iconName="chevron-right"
              />
            </OakFlex>
          </OakFlex>
          {isHighlighted && (
            <VisuallyHidden>&nbsp;(highlighted)</VisuallyHidden>
          )}
        </OakFlex>
      </StyledLink>
    </FocusIndicator>
  );
}
