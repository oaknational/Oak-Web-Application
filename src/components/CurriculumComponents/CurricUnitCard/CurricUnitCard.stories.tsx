import { OakThemeProvider, oakDefaultTheme } from "@oaknational/oak-components";
import { Meta, StoryObj } from "@storybook/nextjs";

import { CurricLessonWarning } from "../CurricLessonWarning";

import Component from "./CurricUnitCard";
import { unitWithOptions, unitWithoutOptions } from "./CurricUnitCard.fixtures";

const meta: Meta<typeof Component> = {
  component: Component,
  argTypes: {},
};

export default meta;
type Story = StoryObj<typeof Component>;

export const CurricUnitCard: Story = {
  args: {
    unit: unitWithOptions,
    index: 10,
    isHighlighted: false,
    href: "#",
  },
  render: function Render(args) {
    return (
      <OakThemeProvider theme={oakDefaultTheme}>
        <Component {...args} />
      </OakThemeProvider>
    );
  },
};

export const CurricUnitCardHighlighted: Story = {
  args: {
    unit: unitWithOptions,
    index: 10,
    isHighlighted: true,
    href: "#",
  },
  render: function Render(args) {
    return (
      <OakThemeProvider theme={oakDefaultTheme}>
        <Component {...args} />
      </OakThemeProvider>
    );
  },
};

export const CurricUnitCardAdditionalSlot: Story = {
  args: {
    unit: unitWithoutOptions,
    index: 10,
    href: "#",
  },
  render: function Render(args) {
    return (
      <OakThemeProvider theme={oakDefaultTheme}>
        <Component
          {...args}
          additional={<CurricLessonWarning count={8} total={10} />}
        />
      </OakThemeProvider>
    );
  },
};
