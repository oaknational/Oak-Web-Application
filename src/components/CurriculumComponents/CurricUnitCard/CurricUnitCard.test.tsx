import CurricUnitCard from "./CurricUnitCard";
import { unitWithOptions } from "./CurricUnitCard.fixtures";

import renderWithTheme from "@/__tests__/__helpers__/renderWithTheme";

describe("CurricUnitCard", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("render", () => {
    const { baseElement } = renderWithTheme(
      <CurricUnitCard
        unit={unitWithOptions}
        index={0}
        isHighlighted={false}
        href="/"
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  test("highlighted state", () => {
    const { baseElement } = renderWithTheme(
      <CurricUnitCard
        unit={unitWithOptions}
        index={0}
        isHighlighted={true}
        href="/"
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  test("index is correct", () => {
    const { baseElement } = renderWithTheme(
      <CurricUnitCard
        unit={unitWithOptions}
        index={0}
        isHighlighted={false}
        href="/"
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });

  test("changes color when highlighed", () => {
    const { baseElement } = renderWithTheme(
      <CurricUnitCard
        unit={unitWithOptions}
        index={0}
        isHighlighted={false}
        href="/"
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });
});
