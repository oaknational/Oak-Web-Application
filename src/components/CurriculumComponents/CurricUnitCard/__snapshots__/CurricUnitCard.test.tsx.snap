// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurricUnitCard changes color when highlighed 1`] = `
<body>
  <div>
    <div
      class="sc-fqkvVR sc-1c55d370-0 jdSkEU iHxKuU"
    >
      <a
        class="sc-c9b8cdc6-0 kcoFyu"
        data-testid="unit-info-link"
        href="/"
      >
        <div
          class="sc-fqkvVR sc-kAyceB hjukdb eeevb"
          data-testid="unit-card"
        >
          <div
            class="sc-fqkvVR sc-kAyceB fgBWrW moWty"
          >
            <div
              class="sc-9d6b6245-0 bZsOFK pa11y-ignore"
            >
              <span
                aria-hidden="true"
              >
                1
              </span>
            </div>
            <span
              class="sc-dAlyuH gYScdJ"
            >
              Something nice
            </span>
            <div
              class="sc-fqkvVR jqPdK"
              data-testid="options-tag"
            >
              <span
                class="sc-dAlyuH nKoIM"
                color="lavender"
              >
                2 unit options
              </span>
            </div>
          </div>
          <div
            class="sc-fqkvVR sc-kAyceB julalA kUbvNw"
          >
            <div
              class="sc-fqkvVR sc-kAyceB fgBWrW gCgXGf"
            >
              <div
                class="sc-fqkvVR sc-fPXMVe kxMTGH wBUcz"
              >
                Unit info
              </div>
              <div
                class="sc-fqkvVR kbnxKi"
              >
                <img
                  alt=""
                  class="sc-dcJsrY beWtwI"
                  data-nimg="fill"
                  decoding="async"
                  loading="lazy"
                  src="https://next_public_oak_assets_host/NEXT_PUBLIC_OAK_ASSETS_PATH/v1707752509/icons/vk9xxxhnsltsickom6q9.svg"
                  style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                />
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</body>
`;

exports[`CurricUnitCard highlighted state 1`] = `
<body>
  <div>
    <div
      class="sc-fqkvVR sc-1c55d370-0 jdSkEU dNUCNM"
    >
      <a
        class="sc-c9b8cdc6-0 fayYhg"
        data-testid="unit-info-link"
        href="/"
      >
        <div
          class="sc-fqkvVR sc-kAyceB jFkkJb eeevb"
          data-testid="highlighted-unit-card"
        >
          <div
            class="sc-fqkvVR sc-kAyceB fgBWrW moWty"
          >
            <div
              class="sc-9d6b6245-0 bZsOFK pa11y-ignore"
            >
              <span
                aria-hidden="true"
              >
                1
              </span>
            </div>
            <span
              class="sc-dAlyuH gYScdJ"
            >
              Something nice
            </span>
            <div
              class="sc-fqkvVR jqPdK"
              data-testid="options-tag"
            >
              <span
                class="sc-dAlyuH nKoIM"
                color="lavender"
              >
                2 unit options
              </span>
            </div>
          </div>
          <div
            class="sc-fqkvVR sc-kAyceB julalA kUbvNw"
          >
            <div
              class="sc-fqkvVR sc-kAyceB fgBWrW gCgXGf"
            >
              <div
                class="sc-fqkvVR sc-fPXMVe kxMTGH wBUcz"
              >
                Unit info
              </div>
              <div
                class="sc-fqkvVR kbnxKi"
              >
                <img
                  alt=""
                  class="sc-dcJsrY bDcxkE"
                  data-nimg="fill"
                  decoding="async"
                  loading="lazy"
                  src="https://next_public_oak_assets_host/NEXT_PUBLIC_OAK_ASSETS_PATH/v1707752509/icons/vk9xxxhnsltsickom6q9.svg"
                  style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                />
              </div>
            </div>
          </div>
          <div
            style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
          >
            Â (highlighted)
          </div>
        </div>
      </a>
    </div>
  </div>
</body>
`;

exports[`CurricUnitCard index is correct 1`] = `
<body>
  <div>
    <div
      class="sc-fqkvVR sc-1c55d370-0 jdSkEU iHxKuU"
    >
      <a
        class="sc-c9b8cdc6-0 kcoFyu"
        data-testid="unit-info-link"
        href="/"
      >
        <div
          class="sc-fqkvVR sc-kAyceB hjukdb eeevb"
          data-testid="unit-card"
        >
          <div
            class="sc-fqkvVR sc-kAyceB fgBWrW moWty"
          >
            <div
              class="sc-9d6b6245-0 bZsOFK pa11y-ignore"
            >
              <span
                aria-hidden="true"
              >
                1
              </span>
            </div>
            <span
              class="sc-dAlyuH gYScdJ"
            >
              Something nice
            </span>
            <div
              class="sc-fqkvVR jqPdK"
              data-testid="options-tag"
            >
              <span
                class="sc-dAlyuH nKoIM"
                color="lavender"
              >
                2 unit options
              </span>
            </div>
          </div>
          <div
            class="sc-fqkvVR sc-kAyceB julalA kUbvNw"
          >
            <div
              class="sc-fqkvVR sc-kAyceB fgBWrW gCgXGf"
            >
              <div
                class="sc-fqkvVR sc-fPXMVe kxMTGH wBUcz"
              >
                Unit info
              </div>
              <div
                class="sc-fqkvVR kbnxKi"
              >
                <img
                  alt=""
                  class="sc-dcJsrY beWtwI"
                  data-nimg="fill"
                  decoding="async"
                  loading="lazy"
                  src="https://next_public_oak_assets_host/NEXT_PUBLIC_OAK_ASSETS_PATH/v1707752509/icons/vk9xxxhnsltsickom6q9.svg"
                  style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                />
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</body>
`;

exports[`CurricUnitCard render 1`] = `
<body>
  <div>
    <div
      class="sc-fqkvVR sc-1c55d370-0 jdSkEU iHxKuU"
    >
      <a
        class="sc-c9b8cdc6-0 kcoFyu"
        data-testid="unit-info-link"
        href="/"
      >
        <div
          class="sc-fqkvVR sc-kAyceB hjukdb eeevb"
          data-testid="unit-card"
        >
          <div
            class="sc-fqkvVR sc-kAyceB fgBWrW moWty"
          >
            <div
              class="sc-9d6b6245-0 bZsOFK pa11y-ignore"
            >
              <span
                aria-hidden="true"
              >
                1
              </span>
            </div>
            <span
              class="sc-dAlyuH gYScdJ"
            >
              Something nice
            </span>
            <div
              class="sc-fqkvVR jqPdK"
              data-testid="options-tag"
            >
              <span
                class="sc-dAlyuH nKoIM"
                color="lavender"
              >
                2 unit options
              </span>
            </div>
          </div>
          <div
            class="sc-fqkvVR sc-kAyceB julalA kUbvNw"
          >
            <div
              class="sc-fqkvVR sc-kAyceB fgBWrW gCgXGf"
            >
              <div
                class="sc-fqkvVR sc-fPXMVe kxMTGH wBUcz"
              >
                Unit info
              </div>
              <div
                class="sc-fqkvVR kbnxKi"
              >
                <img
                  alt=""
                  class="sc-dcJsrY beWtwI"
                  data-nimg="fill"
                  decoding="async"
                  loading="lazy"
                  src="https://next_public_oak_assets_host/NEXT_PUBLIC_OAK_ASSETS_PATH/v1707752509/icons/vk9xxxhnsltsickom6q9.svg"
                  style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                />
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</body>
`;
